Zend/zend_language_parser.y

```diff
@@ -422,21 +429,55 @@ is_variadic:
 ;
 
 class_declaration_statement:
-		class_type { $<num>$ = CG(zend_lineno); }
+		class_modifiers T_CLASS { $<num>$ = CG(zend_lineno); }
 		T_STRING extends_from implements_list backup_doc_comment '{' class_statement_list '}'
-			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, $1, $<num>2, $6,
-				  zend_ast_get_str($3), $4, $5, $8); }
-	|	T_INTERFACE { $<num>$ = CG(zend_lineno); }
-		T_STRING interface_extends_list backup_doc_comment '{' class_statement_list '}'
-			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, ZEND_ACC_INTERFACE, $<num>2, $5,
-				  zend_ast_get_str($3), NULL, $4, $7); }
+			{
+				if (!($1 & ZEND_ACC_PPP_MASK)) { $1 |= ZEND_ACC_PUBLIC; }
+				$$ = zend_ast_create_decl(ZEND_AST_CLASS, $1, $<num>3, $7, zend_ast_get_str($4), $5, $6, $9); 
+			}
+	|	T_CLASS { $<num>$ = CG(zend_lineno); }
+		T_STRING extends_from implements_list backup_doc_comment '{' class_statement_list '}'
+			{ $$ = zend_ast_create_decl(ZEND_AST_CLASS, ZEND_ACC_PUBLIC, $<num>2, $6, zend_ast_get_str($3), $4, $5, $8); }
+;
+
+class_modifiers:
+		class_modifier 					{ $$ = $1; }
+	|	class_modifiers class_modifier 	{ $$ = zend_add_class_modifier($1, $2); }
+;
+
+class_modifier:
+		T_ABSTRACT 				{ $$ = ZEND_ACC_EXPLICIT_ABSTRACT_CLASS; }
+	|	T_FINAL 				{ $$ = ZEND_ACC_FINAL; }
+	|	visibility_modifier 	{ $$ = $1; }
+;
+
+visibility_modifier:
+		T_PUBLIC 		{ $$ = ZEND_ACC_PUBLIC; }
+	|	T_PRIVATE 		{ $$ = ZEND_ACC_PRIVATE; }
```

Comment:
For the first question. Yes, I think it should be valid but pointless :)

I can see the benefit both ways.
* "private for sub-namespaces but public for parents"
* "public for sub-namespaces but private for parents"

I can't really think of a good canonical example for this, but how about:
```php
namespace Project {
    protected interface BackendAdapter {}
}

namespace Project\MyAdapter {
    class Adapter implements \Project\AdapterInterface {}
}
```
If `Project\MyAdapter` is large and has lots of helper classes across many files, it makes sense to put all of the related code in a namespace, but if you have many adapters, but the interface itself does not belong in that more specific namespace.


For the second question, I'm having trouble deciding too. I think I would expect it to be allowed. Ignoring how class extension works internally, the code that actually does the instantiation is defined in correct namespace. If we consider `parent::create` as a method call on the "magic" parent object, I don't think this is _much_ different to the following

```php
namespace Foo {
    interface Processor {}

    private class OneProcessor implements Processor {}

    class ProcessorFactory {
        public function create($type) {
            $className = sprintf("%s\%sProcessor", __NAMESPACE__, ucfirst($type));
            return new $className();
        }
    }
}

namespace Bar {
    class ProcessorFactory {
        private $p;

        public function __construct(\Foo\Processor $p) {
            $this->p = $p
        }

        public function create($type) {
            if ($type !== 'other') {
                return $this->p->create($type); 
            }

            return new OtherProcessor();
        }
    }
}
```